File: c++/data_structure/unionfind.cpp
//ATC001 B
#include <bits/stdc++.h>
using namespace std;

int N, Q;

struct Unionfind {
    vector<int> par, rank;
    
    Unionfind(int n) {
        par.resize(n);
        fill(par.begin(), par.end(), -1);
        rank.resize(n);
    }
    
    int root(int x) {
        int r = x;
        while (!(par[r]<0)) r = par[r];
        int t = x;
        
        while (t!=r) {
            int tmp = t;
            t = par[t];
            par[tmp] = r;
        }
        
        return r;
    }
    
    void unite(int x, int y) {
        int rx = root(x);
        int ry = root(y);
        
        if (rx==ry) return;
        
        if (rank[rx]<=rank[ry]) {
            par[ry] += par[rx];
            par[rx] = ry;
            
            if (rank[rx]==rank[ry]) rank[ry]++;
        }
        else {
            par[rx] += par[ry];
            par[ry] = rx;
        }
    }
    
    bool is_same(int x, int y) {
        return root(x)==root(y);
    }
    
    int count(int x) {
        return -par[root(x)];
    }
};

int main() {
    cin.tie(0); ios::sync_with_stdio(false);
    
    cin >> N >> Q;
    Unionfind uf = Unionfind(N);
    
    for (int q=0; q<Q; q++) {
        int P, A, B; cin >> P >> A >> B;
        
        if (P==0) uf.unite(A, B);
        else {
            if (uf.is_same(A, B)) cout << "Yes" << endl;
            else cout << "No" << endl;
        }
    }
}

File: python/data_structure/unionfind.py
#ATC001 B
import sys
input = sys.stdin.readline

class Unionfind:
    def __init__(self, n):
        self.par = [-1]*n
        self.rank = [1]*n
    
    def root(self, x):
        r = x
        
        while not self.par[r]<0:
            r = self.par[r]
        
        t = x
        
        while t!=r:
            tmp = t
            t = self.par[t]
            self.par[tmp] = r
        
        return r
    
    def unite(self, x, y):
        rx = self.root(x)
        ry = self.root(y)
        
        if rx==ry:
            return
        
        if self.rank[rx]<=self.rank[ry]:
            self.par[ry] += self.par[rx]
            self.par[rx] = ry
            
            if self.rank[rx]==self.rank[ry]:
                self.rank[ry] += 1
        else:
            self.par[rx] += self.par[ry]
            self.par[ry] = rx
    
    def is_same(self, x, y):
        return self.root(x)==self.root(y)
    
    def count(self, x):
        return -self.par[self.root(x)]
        
N, Q = map(int, input().split())
uf = Unionfind(N)

for _ in range(Q):
    P, A, B = map(int, input().split())
    
    if P==0:
        uf.unite(A, B)
    else:
        if uf.is_same(A, B):
            print('Yes')
        else:
            print('No')
